tokio-postgres/src/connect.rs

```diff
@@ -15,35 +15,67 @@ pub async fn connect<T>(
 where
     T: MakeTlsConnect<Socket>,
 {
-    if config.host.is_empty() {
-        return Err(Error::config("host missing".into()));
+    if config.host.is_empty() && config.hostaddr.is_empty() {
+        return Err(Error::config("both host and hostaddr are missing".into()));
     }
 
-    if config.port.len() > 1 && config.port.len() != config.host.len() {
+    if !config.host.is_empty()
+        && !config.hostaddr.is_empty()
+        && config.host.len() != config.hostaddr.len()
+    {
+        let msg = format!(
+            "number of hosts ({}) is different from number of hostaddrs ({})",
+            config.host.len(),
+            config.hostaddr.len(),
+        );
+        return Err(Error::config(msg.into()));
+    }
+
+    // At this point, either one of the following two scenarios could happen:
+    // (1) either config.host or config.hostaddr must be empty;
+    // (2) if both config.host and config.hostaddr are NOT empty; their lengths must be equal.
+    let num_hosts = cmp::max(config.host.len(), config.hostaddr.len());
+
+    if config.port.len() > 1 && config.port.len() != num_hosts {
         return Err(Error::config("invalid number of ports".into()));
     }
 
     let mut error = None;
-    for (i, host) in config.host.iter().enumerate() {
+    for i in 0..num_hosts {
+        let host = config.host.get(i);
+        let hostaddr = config.hostaddr.get(i);
         let port = config
             .port
             .get(i)
             .or_else(|| config.port.first())
             .copied()
             .unwrap_or(5432);
 
+        // The value of host is always used as the hostname for TLS validation.
+        // postgres doesn't support TLS over unix sockets, so the choice for Host::Unix variant here doesn't matter
         let hostname = match host {
-            Host::Tcp(host) => host.as_str(),
-            // postgres doesn't support TLS over unix sockets, so the choice here doesn't matter
-            #[cfg(unix)]
-            Host::Unix(_) => "",
+            Some(Host::Tcp(host)) => host.as_str(),
```

Comment:
This is missing the `None` case. It should probably just stringify the hostaddr in that case.

Reply:
Thanks! Sounds reasonable. Originally, I wanted to just fallback to empty string for both Unix and None case. It's better to just use the `hostaddr` as hostname if present. TLS verification is likely going to fail anyway, but still better than an empty string. Fixed in latest commit, thank you!